Games/2048/main.cpp

```diff
@@ -131,7 +131,7 @@ int main(int argc, char** argv)
 
         GUI::MessageBox::show(window, "New settings have been set and will be applied on the next game", "Settings Changed Successfully", GUI::MessageBox::Type::Information);
     };
-    auto start_a_new_game = [&] {
+    auto start_a_new_game = [&, update] {
```

Comment:
How is this thing captured here if not by ref?
lambdas with captures shouldn't be copyable, so that's either a const reference or just a plain reference?

Reply:
Yeah no, it's grabbed by const reference, even when `start_a_new_game` is explicitly mutable:
```c++
auto start_a_new_game = [&, update]() mutable {
        update = 1;
```

->

```
../Games/2048/main.cpp:101:21: note:   no known conversion for argument 1 from
'int' to 'const main(int, char**)::<lambda()>&'
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Interesting! In FontEditor it's clearly a `const` value, not ref:

```
../Applications/FontEditor/FontEditor.cpp:267:23: error: no match for 'operator=' (operand types are 'const FontEditorWidget::FontEditorWidget(const AK::String&, AK::RefPtr<Gfx::Font>&&)::<lambda()>' and 'int')
  267 |         update_demo = 1;
      |                       ^
```

Why can it be copied here, but not there?

Reply:
Lambdas *are* copyable. That's what `TextEditor: Don't try to move(lambda)` relies on, and here it's in the specification, ctrl-f for "copy constructor": https://en.cppreference.com/w/cpp/language/lambda

Reply:
Did you mark the containing lambda as `mutable`?
Come think of it I'm not exactly sure how it determines if a lambda is copyable.
(If it's just "if all captured refs are copyable", then no `GUI::Widget` is copyable, and so it shouldn't even compile...)

Reply:
> Lambdas are copyable

IIRC only ones with no captures?

Reply:
Well, it says "according to the usual rules for copy constructors", so:
- In FontEditor, `update_demo` stores a reference to some labels. References are copyable. Therefore, `update_demo` is copyable.
- In 2048, `update` takes everything by reference. Therefore, `update` is copyable.

So … no problem there, right?

Reply:
I suppose that makes sense.
Though copying lambdas feels weird, but eh, better than it being wrong

Reply:
The same copy/move requirements should apply to lambdas and their captures. A lambda is copyable as long as all captures are, and movabale as long as all captures are. references/const references are always copyable/movable.

Reply:
Yup, that's why I've been quoting the exact same thing from the [reference](https://en.cppreference.com/w/cpp/language/lambda) :)

> better than it being wrong

That's a critical point actually, because most of usages are *correct* – but error-prone, as we have now seen with FontEditor, and that's why I'm making all these changes.

So I'd say "better copy a few times too many (and waste a hundred bytes) than once too few (and try to read from an invalid stackframe)".

What do you think about that? If you don't like it, I can just chuck all the other changes, and only keep the rest (i.e., fixes, explanations, and lambda-removals).

Reply:
In this case, `update` should be not copied. An implicit reference to the value on the stack is perfectly ok. Besides, `update` itself references the stack. Copying it probably won't hurt anything, but it won't solve any problem. It's fine the way it is right now. None of the stack variables ever go out of scope until the main function exits, so it's perfectly fine to do it the way it's done currently.

Reply:
That actually answers my questions. Alright, I'll chuck the "precautionary" commits :)

Reply:
I removed the commits in question, which resolves this discussion as far as I understand. If you disagree, or if I misunderstood something, please respond :)